lib/recursive-methods.rb

```diff
@@ -1,49 +1,91 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the value of the input
+# Space complexity: O(1) because there's nothing storing the result
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+  return 1 if n == 0
+  return 1 if n == 1
+  raise ArgumentError if n < 0
+  return n * factorial(n - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n) where n is the length of s
+# Space complexity: O(n) where n is the length of s
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+  return s if s.length == 0
+  last = s[-1]
+  s = s[0..-2]
```

Comment:
This creates a new string which increases the time and space time complexity.

So the time and space complexity is `O(n^2)`.